be/src/exec/pipeline/jdbc_scan_operator.cpp

```diff
@@ -0,0 +1,223 @@
+    std::string driver_checksum = jdbc_table->jdbc_driver_checksum();
+    std::string driver_class = jdbc_table->jdbc_driver_class();
+    std::string driver_location;
+
+    if (status = JDBCDriverManager::getInstance()->get_driver_location(driver_name, driver_url, driver_checksum, &driver_location); !status.ok()) {
+        LOG(ERROR) << fmt::format("Get JDBC Driver[{}] error, error is {}", driver_name, status.to_string());
+        _set_scanner_state(true, status);
+        return;
+    }
+
+    scan_ctx.driver_path = driver_location;
+    scan_ctx.driver_class_name = driver_class;
+    scan_ctx.jdbc_url = jdbc_table->jdbc_url();
+    scan_ctx.user = jdbc_table->jdbc_user();
+    scan_ctx.passwd = jdbc_table->jdbc_passwd();
+    scan_ctx.sql = get_jdbc_sql(jdbc_table->jdbc_table(), _jdbc_scan_node.columns, _jdbc_scan_node.filters, _limit);
+
+    _scanner.reset(new vectorized::JDBCScanner(scan_ctx, _result_tuple_desc));
+
+    if (status = _scanner->open(state); !status.ok()) {
+        _set_scanner_state(true, status);
+        _scanner->close(state);
+        return;
+    }
+    if (status = _fetch_chunks(); !status.ok()) {
+        _set_scanner_state(true, status);
+        _scanner->close(state);
+        return;
+    }
+    _scanner->close(state);
+    _set_scanner_state(true, Status::OK());
+}
+
+void JDBCScanOperator::set_finishing(RuntimeState* state) {
+    _is_finished.store(true);
+}
+
+StatusOr<vectorized::ChunkPtr> JDBCScanOperator::pull_chunk(RuntimeState* state) {
+    if (_is_finished.load()) {
+        return nullptr;
+    }
+    Status scanner_status;
+    bool scanner_finished = _is_scanner_finished(&scanner_status);
+    if (scanner_finished && !scanner_status.ok()) {
+        return scanner_status;
+    }
+    ChunkPtr chunk;
+    int ret = _result_chunks.try_get(&chunk);
+    switch (ret) {
+    case -1: {
+        return nullptr;
+    }
+    case 0: {
+        // no new chunks and scanner thread is finished
+        if (scanner_finished) {
+            _is_finished.store(true);
+        }
+        return nullptr;
+    }
+    case 1: {
+        return chunk;
+    }
+    default: {
+        break;
+    }
+    }
+    DCHECK(false) << "unreachable path";
+    return nullptr;
+}
+
+void JDBCScanOperator::_set_scanner_state(bool is_finished, const Status& new_status) {
+    std::lock_guard<SpinLock> l(_scanner_state_mutex);
+    _scanner_finished = is_finished;
+    _scanner_status = new_status;
+}
+
+bool JDBCScanOperator::_is_scanner_finished(Status* status) const {
+    std::lock_guard<SpinLock> l(_scanner_state_mutex);
+    if (status) {
+        *status = _scanner_status;
+    }
+    return _scanner_finished;
+}
+
+Status JDBCScanOperator::_fetch_chunks() {
+    bool eof = false;
+    while (!_is_finished.load() && !eof) {
+        ChunkPtr chunk;
+        RETURN_IF_ERROR(_scanner->get_next(_state, &chunk, &eof));
+        if (!_result_chunks.blocking_put(chunk)) {
+            break;
+        }
```

Comment:
`blocking_put` returns false means what? Unexpected error?

Reply:
> `blocking_put` returns false means what? Unexpected error?

Only if `_result_chunks` shutdown, it will return false. But this scenario will not occur in our case, because the exit is controlled by `_is_finished`, we won't invoke `shutdown` method on `_result_chunks`